Server/src/main/java/org/openas2/lib/helper/BCCryptoHelper.java

```diff
@@ -396,10 +386,31 @@ public MimeBodyPart verifySignature(MimeBodyPart part, Certificate cert) throws
                     logger.trace("Signer Attributes: data not available.");
                 }
             }
-            if (signer.verify(signerInfoVerifier)) {
-                logSignerInfo("Verified signature for signer info", signer, part, x509Cert);
-                return signedPart.getContent();
+
+            // Claudio Degioanni claudio.degioanni@bmeweb.it 05/11/2021
+            try {
+                String as2SignIgnoreTimeIssue = Properties.getProperty("as2_sign_allow_expired_certificate", "false");
+
+                if ("false".equalsIgnoreCase(as2SignIgnoreTimeIssue)) {
+                    // normal check
+                    if (signer.verify(signerInfoVerifier)) {
+                        logSignerInfo("Verified signature for signer info", signer, part, x509Cert);
+                        return signedPart.getContent();
+                    }
+                } else {
+                    signer = new SignerInfoIgnoringExpiredCertificate(signer);
+                    // if flag is enabled log only issue
+                    if (signer.verify(signerInfoVerifier)) {
+                        logSignerWarn("Verified signature for signer info EXCLUDING certificate date verification (OUTDATED CERTIFICATE)", signer, part, x509Cert);
+                        return signedPart.getContent();
+                    }
+                }
+
+            } catch (CMSVerifierCertificateNotValidException ex) {
+                throw ex;
             }
```

Comment:
Can you change the check so that it only occuirs inside the catch block. That way it is not necessary to change the configuration whe the certificate is updated to have valid dates:

```
            try {
                String as2SignIgnoreTimeIssue = Properties.getProperty("as2_sign_allow_expired_certificate", "false");
                // normal check
                if (signer.verify(signerInfoVerifier)) {
                    logSignerInfo("Verified signature for signer info", signer, part, x509Cert);
                    return signedPart.getContent();
                }
            } catch (CMSVerifierCertificateNotValidException ex) {
                if ("true".equalsIgnoreCase(as2SignIgnoreTimeIssue)) {
                    signer = new SignerInfoIgnoringExpiredCertificate(signer);
                    // if flag is enabled log only issue
                    if (signer.verify(signerInfoVerifier)) {
                        logSignerWarn("Verified signature for signer info EXCLUDING certificate date verification (OUTDATED CERTIFICATE)", signer, part, x509Cert);
                        return signedPart.getContent();
                    }
                }
                } else {
                    throw ex;
                }
            }
```

Reply:
I have changed the code as requested